### ISSUE TITLE
Download but only start new release on X

### ISSUE THREAD

**Issue by 20k-ultra** [balena-io/balena-supervisor#1624]
Having an update strategy that only applies at X condition would allow for more robust release deployments. This issue is pretty generic because what would X be ? This list can be of infinite size...

To start we are tracking the following conditions:

 1. reboot

**Comment by codewithcheese** [comment]
Update locks and overrides seems like they could be replaced by this if X is time. Currently lock serves as "do not update immediately" and the override serving as "ok update now".  Users could set X to be a range of time that is appropriate for sensitive updates, or they can leave it unspecified. The equivalent to an override would be to set X = next 30 minutes.

From discussion here: https://www.flowdock.com/app/rulemotion/r-supervisor/threads/BeFK56X0dcB3JyikGrKwSHWWzfb

**Comment by codethief** [comment]
I came here to suggest this feature because in our project we cannot restart the application (or do a hand-over) during runtime without sacrificing user experience. What we would prefer is: The supervisor downloads the update while keeping the old version running. Once the download has finished, the supervisor adds flags to the new and the old version, so that after rebooting the device the new version gets started and the old version gets deleted. The user can then decide on their own when to reboot the device.

**Comment by 20k-ultra** [comment]
It sounds like you want "Download but only start new release on reboot" @codethief ?

**Comment by codethief** [comment]
@20k-ultra Yes, correct. Is this already possible? I didn't find it in the [docs](https://www.balena.io/docs/learn/deploy/release-strategy/update-strategies/).

**Comment by 20k-ultra** [comment]
we currently don't have this type of update strategy and I can't think of a workaround right now that you could use. I'll see how we can prioritize this issue.

**Comment by 20k-ultra** [comment]
I've just thought of an update strategy which allows you to download but only run on reboot and many more conditions (infinite).

If we made a **download-then-wait** strategy, the sequence of events would be:

    1. create /tmp/install.lock
    2. download release images
    3. check if install lock is gone
       3a. Lock exists then do nothing (log that we are waiting for install lock to go away)
    4. Install containers (lock does not exist)

Because this lock exists in the tmp folder then on reboot it will be removed and allow new containers to be created. Your old containers will start though if you have a `restart: always` policy so to prevent that you must use a `restart: no` policy. Otherwise, users can also have the same behaviour without a reboot by removing the install.lock from their application. Their application can then perform the condition checking to know when to remove the install.lock.

**Comment by cywang117** [comment]
This may fit with the delta/image update preload method we tested out recently, which may work for both upgrades and rollbacks. We could incorporate this delta preload method as part of this Supervisor issue.

**Comment by maggie0002** [comment]
> I've just thought of an update strategy which allows you to download but only run on reboot and many more conditions (infinite).
> 
> If we made a **download-then-wait** strategy, the sequence of events would be:
> 
> ```
> 1. create /tmp/install.lock
> 2. download release images
> 3. check if install lock is gone
>    3a. Lock exists then do nothing (log that we are waiting for install lock to go away)
> 4. Install containers (lock does not exist)
> ```
> 
> Because this lock exists in the tmp folder then on reboot it will be removed and allow new containers to be created. Your old containers will start though if you have a `restart: always` policy so to prevent that you must use a `restart: no` policy. Otherwise, users can also have the same behaviour without a reboot by removing the install.lock from their application. Their application can then perform the condition checking to know when to remove the install.lock.

This sounds exactly the same as the current update lock mechanism: https://www.balena.io/docs/learn/deploy/release-strategy/update-locking/. 

> The presence of a lockfile will ensure that your application does not get killed, but updates will still be downloaded by the supervisor, ready to be applied once the lockfile no longer exists.

It would be the equivalent of attaching that mechanism to a 'download-then-wait' option. Although my experience using it is the constant echoing in the log is a pain as it floods the window over time and hides other relevant messages.
